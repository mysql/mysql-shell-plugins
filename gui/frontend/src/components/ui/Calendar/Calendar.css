/*
 * Copyright (c) 2020, 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

.msg.calendarHost,
.msg.calendarHost::before {
    background-color: var(--window-background);
}

.msg.calendarView {
    width: 250px;
    height: 250px;
    grid-template-rows: repeat(8, 1fr);
}

.msg.calendarView .header {
    background-color: var(--window-headerBackground);
}

.msg.calendarView .gridCell {
    justify-content: center;
    align-items: center;
}

.msg.calendarView #title {
    font-size: 1.1rem;
    font-weight: 800;
}

.msg.calendarView .button {
    min-width: 16px;
    width: 22px;
    height: 22px;
    border: none;
    background: none;
}

.msg.calendarView #prevMonthButton::after {
    display: inline-block;
    content: "";
    width: 7px;
    height: 7px;

    position: absolute;

    border-right: 1px solid var(--foreground);
    border-bottom: 1px solid var(--foreground);

    transform: translate(2px, 0px) rotate(135deg);
    pointer-events: none;
}

.msg.calendarView #nextMonthButton::after {
    display: inline-block;
    content: "";
    width: 7px;
    height: 7px;

    position: absolute;

    border-right: 1px solid var(--foreground);
    border-bottom: 1px solid var(--foreground);

    transform: translate(-2px, 0px) rotate(-45deg);
    pointer-events: none;
}
