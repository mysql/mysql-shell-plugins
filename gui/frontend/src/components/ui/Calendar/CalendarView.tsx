/*
 * Copyright (c) 2020, 2023, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { ComponentChild } from "preact";

import { Button } from "../Button/Button.js";
import { IComponentProperties, ComponentBase } from "../Component/ComponentBase.js";
import { ContentAlignment } from "../Container/Container.js";
import { Grid } from "../Grid/Grid.js";
import { GridCell } from "../Grid/GridCell.js";
import { Label } from "../Label/Label.js";

interface ICalendarViewProperties extends IComponentProperties {
    date: Date;

    onSelect?: (date: Date, props: ICalendarViewProperties) => void;
    onChangeMonth?: (inc: boolean) => void;
}

export class CalendarView extends ComponentBase<ICalendarViewProperties> {

    private dayNames: string[] = [];

    public constructor(props: ICalendarViewProperties) {
        super(props);

        // Initialize localized day names here. We need a concrete date for that.
        for (let day = 19; day < 26; ++day) {
            const date = new Date(2020, 3, day);
            this.dayNames.push(date.toLocaleString("default", { weekday: "short" }));
        }
    }

    public render(): ComponentChild {
        const { date } = this.mergedProps;
        const className = this.getEffectiveClassNames(["calendarView"]);

        // View caption (the current month + year).
        const month = date.toLocaleString("default", { month: "long" });
        const monthYear = `${month} ${date.getFullYear()}`;

        // Weekdays.
        const dayNames = [];
        let name = "";
        for (let i = 1; i < 7; ++i) { // Monday through Saturday.
            name = this.dayNames[i];
            dayNames.push(
                <GridCell
                    key={name}
                    className="header"
                >
                    {name}
                </GridCell>,
            );
        }

        name = this.dayNames[0]; // Sunday.
        dayNames.push(
            <GridCell
                key={name}
                className="header"
            >
                {name}
            </GridCell>,
        );

        // Day numbers.
        // First figure out the weekday this month starts with and the number of days in this month.
        const firstOfMonth = new Date(date.getFullYear(), date.getMonth(), 0);
        const lastOfMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0);
        const startDay = firstOfMonth.getDay();
        const dayCount = lastOfMonth.getDate();

        // Add empty cells for days that do not belong to this month.
        const days = [];
        for (let i = 0; i < startDay; ++i) {
            days.push(<GridCell key={`preDay${i}`} />);
        }

        // Add the actual day numbers.
        for (let i = 1; i <= dayCount; ++i) {
            days.push(
                <GridCell key={`day${i}`}>
                    <Button
                        caption={i.toString()}
                        onClick={this.handleDayClick}
                    />
                </GridCell>,
            );
        }

        // To ensure we always have 6 grid rows add more empty cells.
        while (days.length < 36) {
            days.push(<GridCell key={`preDay${days.length}`} />);
        }

        return (
            <Grid
                className={className}
                columns={7}
            >
                <GridCell
                    className="header"
                >
                    <Button
                        id="prevMonthButton"
                        onClick={this.handleMonthSwitch}
                    />
                </GridCell>
                <GridCell
                    columnSpan={5}
                    crossAlignment={ContentAlignment.Center}
                    className="header"
                >
                    <Label id="title" caption={monthYear} />
                </GridCell>
                <GridCell
                    className="header"
                >
                    <Button
                        id="nextMonthButton"
                        onClick={this.handleMonthSwitch}
                    />
                </GridCell>
                {dayNames}
                {days}
            </Grid>
        );
    }

    private handleMonthSwitch = (e: MouseEvent | KeyboardEvent, props: IComponentProperties): void => {
        const { onChangeMonth } = this.mergedProps;
        onChangeMonth?.(props.id === "nextMonthButton");
    };

    private handleDayClick = (e: MouseEvent | KeyboardEvent): void => {
        const day = parseInt((e.target as HTMLElement).innerText, 10);
        const { date, onSelect } = this.mergedProps;
        onSelect?.(new Date(date.getFullYear(), date.getMonth(), day), this.mergedProps);
    };

}
