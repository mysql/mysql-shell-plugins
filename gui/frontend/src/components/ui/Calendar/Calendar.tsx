/*
 * Copyright (c) 2020, 2024, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import "./Calendar.css";

import { ComponentChild, createRef } from "preact";

import { CalendarView } from "./CalendarView.js";
import {
    IComponentProperties, IComponentState, ComponentBase, ComponentPlacement,
} from "../Component/ComponentBase.js";
import { Popup } from "../Popup/Popup.js";

interface ICalendarProperties extends IComponentProperties {
    initialDate?: Date;

    onChange?: (date: Date) => void;
}

interface ICalendarState extends IComponentState {
    currentDate: Date; // XXX: convert to controlled component.
}

export class Calendar extends ComponentBase<ICalendarProperties, ICalendarState> {

    public static defaultProps = {
        initialDate: new Date(),
    };

    private popupRef = createRef<Popup>();

    public constructor(props: ICalendarProperties) {
        super(props);

        this.state = {
            currentDate: props.initialDate!,
        };

        this.addHandledProperties("initialDate");
    }

    public render(): ComponentChild {
        const { currentDate } = this.state;
        const className = this.getEffectiveClassNames(["calendarHost"]);

        return (
            <Popup
                ref={this.popupRef}
                className={className}
                placement={ComponentPlacement.BottomCenter}
                showArrow={true}
            >
                <CalendarView
                    date={currentDate}
                    onChangeMonth={this.handleMonthChange}
                    onSelect={this.handleSelectDate}
                />
            </Popup>
        );
    }

    public open(currentTarget: HTMLElement): void {
        this.popupRef?.current?.open(currentTarget.getBoundingClientRect());
    }

    public close(): void {
        this.popupRef?.current?.close(false);
    }

    private handleMonthChange = (inc: boolean): void => {
        const { currentDate } = this.state;
        const offset = inc ? 1 : -1;
        this.setState({
            currentDate: new Date(currentDate.getFullYear(), currentDate.getMonth() + offset, currentDate.getDate()),
        });
    };

    private handleSelectDate = (date: Date): void => {
        this.popupRef.current?.close(false);

        const { onChange } = this.mergedProps;
        onChange?.(date);
    };

}
