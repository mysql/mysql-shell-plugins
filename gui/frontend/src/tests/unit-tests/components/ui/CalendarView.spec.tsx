/*
 * Copyright (c) 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mount } from "enzyme";
import React from "react";
import { CalendarView } from "../../../../components/ui/Calendar/CalendarView";
import { snapshotFromWrapper } from "../../test-helpers";


describe("CalendarView testing", () => {

    it("Calendar view test properties", () => {
        const currentDate = new Date("2022-05-02");
        const component = mount(
            <CalendarView
                date={currentDate}
                onChangeMonth={jest.fn}
                onSelect={jest.fn}
            />,
        );
        expect(component).toBeTruthy();
        const props = component.props();
        expect(props.date).toEqual(currentDate);
    });

    it("Render test", () => {
        const currentDate = new Date("2022-05-02");
        const component = mount(
            <CalendarView
                date={currentDate}
                onChangeMonth={jest.fn}
                onSelect={jest.fn}
            />,
        );
        expect(snapshotFromWrapper(component)).toMatchSnapshot();
        component.unmount();
    });

});
