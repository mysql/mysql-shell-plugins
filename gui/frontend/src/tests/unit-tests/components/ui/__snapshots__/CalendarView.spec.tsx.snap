// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalendarView testing > Calendar view snapshot 1`] = `
<div>
  <div
    class="msg grid calendarView"
    style="grid-template-columns: repeat(7, 1fr);"
  >
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      <div
        class="msg button"
        id="prevMonthButton"
        role="button"
        tabindex="0"
      />
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row; align-items: center; grid-column: span 5;"
    >
      <label
        class="msg label"
        data-tooltip="expand"
        id="title"
      >
        May 2022
      </label>
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      <div
        class="msg button"
        id="nextMonthButton"
        role="button"
        tabindex="0"
      />
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      Mon
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      Tue
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      Wed
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      Thu
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      Fri
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      Sat
    </div>
    <div
      class="msg container fixedScrollbar gridCell header"
      style="flex-direction: row;"
    >
      Sun
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    />
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    />
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    />
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    />
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    />
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    />
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="1"
        class="msg button"
        role="button"
        tabindex="0"
      >
        1
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="2"
        class="msg button"
        role="button"
        tabindex="0"
      >
        2
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="3"
        class="msg button"
        role="button"
        tabindex="0"
      >
        3
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="4"
        class="msg button"
        role="button"
        tabindex="0"
      >
        4
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="5"
        class="msg button"
        role="button"
        tabindex="0"
      >
        5
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="6"
        class="msg button"
        role="button"
        tabindex="0"
      >
        6
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="7"
        class="msg button"
        role="button"
        tabindex="0"
      >
        7
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="8"
        class="msg button"
        role="button"
        tabindex="0"
      >
        8
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="9"
        class="msg button"
        role="button"
        tabindex="0"
      >
        9
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="10"
        class="msg button"
        role="button"
        tabindex="0"
      >
        10
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="11"
        class="msg button"
        role="button"
        tabindex="0"
      >
        11
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="12"
        class="msg button"
        role="button"
        tabindex="0"
      >
        12
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="13"
        class="msg button"
        role="button"
        tabindex="0"
      >
        13
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="14"
        class="msg button"
        role="button"
        tabindex="0"
      >
        14
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="15"
        class="msg button"
        role="button"
        tabindex="0"
      >
        15
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="16"
        class="msg button"
        role="button"
        tabindex="0"
      >
        16
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="17"
        class="msg button"
        role="button"
        tabindex="0"
      >
        17
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="18"
        class="msg button"
        role="button"
        tabindex="0"
      >
        18
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="19"
        class="msg button"
        role="button"
        tabindex="0"
      >
        19
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="20"
        class="msg button"
        role="button"
        tabindex="0"
      >
        20
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="21"
        class="msg button"
        role="button"
        tabindex="0"
      >
        21
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="22"
        class="msg button"
        role="button"
        tabindex="0"
      >
        22
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="23"
        class="msg button"
        role="button"
        tabindex="0"
      >
        23
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="24"
        class="msg button"
        role="button"
        tabindex="0"
      >
        24
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="25"
        class="msg button"
        role="button"
        tabindex="0"
      >
        25
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="26"
        class="msg button"
        role="button"
        tabindex="0"
      >
        26
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="27"
        class="msg button"
        role="button"
        tabindex="0"
      >
        27
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="28"
        class="msg button"
        role="button"
        tabindex="0"
      >
        28
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="29"
        class="msg button"
        role="button"
        tabindex="0"
      >
        29
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="30"
        class="msg button"
        role="button"
        tabindex="0"
      >
        30
      </div>
    </div>
    <div
      class="msg container fixedScrollbar gridCell"
      style="flex-direction: row;"
    >
      <div
        caption="31"
        class="msg button"
        role="button"
        tabindex="0"
      >
        31
      </div>
    </div>
  </div>
</div>
`;
